### ISSUE TITLE
More aggressive parallelism in build order?

### ISSUE THREAD

**Issue by mikepurvis** [catkin/catkin_tools#372]
Presently catkin_tools hitches off of `topological_order_packages` from catkin_pkg, to determine the build order. However, this means there are periods of building, particularly with large workspaces, where there are fewer things that can be built than there are cores.

Is there a way to prioritize building packages that unblock other packages, over building leaf packages? Then leaf packages could be "saved up" and worked on when you're otherwise at a choke-point (building roscpp, rviz, etc).

I don't have a clear picture how this would be implemented, but I think the basic idea would be to tag each package with the number of in-workspace packages which depend on that package, and then every time an executor frees up, it sorts the candidates for what it could work on by that number and chooses the largest.

**Comment by jbohren** [comment]
> Is there a way to prioritize building packages that unblock other packages, over building leaf packages? 

So we could order the packages according to recursive dependency "weight" subject to the topological constraint. That way, for any set of queued packages, the packages which have more dependents would be built before the others. As more packages get queued, they will take priority over the leaf packages until all of the queued packages are leaves. Is that what you have in mind?

**Comment by jbohren** [comment]
We could precompute the job dependency weight during initialization, and then just keep the list of queued jobs sorted when we update it here: https://github.com/catkin/catkin_tools/blob/master/catkin_tools/execution/executor.py#L344-L348

**Comment by mikepurvis** [comment]
Yes, that's the idea exactly. Upon further consideration, yes it makes perfect sense that it would be possible to precompute; that's much better than trying to do it live.

**Comment by mikepurvis** [comment]
Given that catkin_tools uses `topological_order_packages`, I think the fix for this would have to go into catkin_pkg:

https://github.com/ros-infrastructure/catkin_pkg/blob/6d79544f7fa4b646f7574910d085806de393bb46/src/catkin_pkg/topological_order.py#L117
